### DISCUSSION TITLE
How do I manually configure Shards for execution

### DISCUSSION THREAD

**Discussion by mvgiacomello** [vitest-dev/vitest#5480] [Q&A] [answered]
In a project I am working on, we use vitest to drive Puppeteer tests. These tests take a lot of resources to run and we know which ones can be run in parallel and which can only run alone in our CI. Is there a way to manually configure which tests run in parallels and which in sequence? Basically manually defining the shards.

We tried the following without much success :
```
// Import tests ABCDEF which are async functions with `test()` and `test.concurrent()` inside.

suite('Test Suite', async () => {
  describe.concurrent('Group A', async () => {
    describe('testA', testA)
    describe('testB', testB)
    describe('testC, testC)
  })
  describe('testD', testD)
  describe.concurrent('Group B', async () => {
    describe('testE', testE)
    describe('testF', testF)
  })
})
```

We're also using the configs:
```
minWorkers: 1
maxWorkers: 2
```

**Comment by hi-ogawa** [comment]
One potential confusion is that `test.concurrenty` is configured by `maxConcurrency` https://vitest.dev/config/#maxconcurrency and `minWorkers/maxWorkers` is for configuring test-file based concurrency.

For example, if you have multiple test files, which should run one by one, and then in each file, you want to use `test.concurrency` but concurrent up to 2, then probably this config would do it:
- `minWorkers = maxWorkers = 1`
- `maxConcurrency = 2`

---

Btw, you didn't mention in which way it's not working. Is it too serial? or is it too concurrent?

> We tried the following without much success :

**Reply by mvgiacomello** [reply]
Humm, thanks @hi-ogawa - okay, so I was missing the `maxConcurrency` in my config file. That should solve part of my problems.

> Btw, you didn't mention in which way it's not working

Going back to my main question. I'd like to set `maxWorkers=2` so I could run 2 files in parallel, however, I'd like to specifically tell `vitest` which files are okay to run in parallel as some tests are quite heavy to run together.

**Reply by hi-ogawa** [reply, marked as answer]
> I'd like to set `maxWorkers=2` so I could run 2 files in parallel, however, I'd like to specifically tell vitest which files are okay to run in parallel as some tests are quite heavy to run together.

I don't think that's currently possible with a single `vitest `command. 
Probably you would need to split to two test commands, something like:

```ts
{
  "test:light": "vitest --exclude '**/*.heavy.test.ts' --maxWorkers 2"
  "test:heavy": "vitest '.heavy.test.ts' --maxWorkers 1"
}
```

Or these two patterns are distinct enough, then you could prepare two configs:

```ts
{
  "test:light": "vitest -c vitest.config.light.ts"
  "test:heavy": "vitest -c vitest.config.heavy.ts"
}
```

**Reply by mvgiacomello** [reply]
Thank you very much, @hi-ogawa, the first suggestion did the job! 👏
